File: javapatterns/src/main/java/org/iut/nantes/state/canards/etats/DansLair.java
package org.iut.nantes.state.canards.etats;

import org.iut.nantes.state.canards.Canard;

public class DansLair implements Etat {
    private Canard canard;

    public DansLair (Canard canard) {
        this.canard = canard;
    }

    @Override
    public void avancer() {
        System.out.println("Je vole!");
    }

    @Override
    public void plongeon() {
        canard.changeEtatSousLeau();
    }

    @Override
    public void trempette() {
        canard.changeEtatSurLeau();
    }

    @Override
    public void decollage() {

    }

    @Override
    public void palmeATerre() {

    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/state/canards/etats/DansLair.kt
package org.iut.nantes.state.canards.etats

import org.iut.nantes.state.canards.Canard

class DansLair(private val canard: Canard) : Etat {
    override fun avancer() {
        println("Je vole!")
    }

    override fun plongeon() {
        canard.changeEtatDansEau()
    }

    override fun trempette() {
        canard.changeEtatSurEau()
    }

    override fun decollage() {
        println("Je vole déjà")
    }

    override fun palmeATerre() { }
}